lib/queue_classic_plus/worker.rb

```diff
@@ -32,30 +32,45 @@ def handle_failure(job, e)
         # We definitely want to retry because the connection was lost mid-task.
         force_retry = true
       end
+
       klass = job_klass(job)
 
       if force_retry && !(klass.respond_to?(:disable_retries) && klass.disable_retries)
-        klass_retries = klass.respond_to?(:max_retries) ? klass.max_retries : 0
-        klass.restart_in(0, (job[:remaining_retries] || klass_retries || 0).to_i, *job[:args])
+        Metrics.increment("qc.force_retry", source: @q_name)
+
+        retry_with_remaining(klass: klass, job: job, e: e, backoff: 0)
       # The mailers doesn't have a retries_on?
       elsif klass && klass.respond_to?(:retries_on?) && klass.retries_on?(e)
-        remaining_retries = (job[:remaining_retries] || klass.max_retries).to_i
-        remaining_retries -= 1
-
-        if remaining_retries > 0
-          klass.restart_in((klass.max_retries - remaining_retries) * BACKOFF_WIDTH,
-                           remaining_retries,
-                           *job[:args])
-        else
-          enqueue_failed(job, e)
-        end
+        Metrics.increment("qc.retry", source: @q_name)
+
+        backoff = (max_retries(klass) - remaining_retries(klass, job)) * BACKOFF_WIDTH
+
+        retry_with_remaining(klass: klass, job: job, e: e, backoff: backoff)
       else
         enqueue_failed(job, e)
       end
 
       FailedQueue.delete(job[:id])
     end
 
+    def retry_with_remaining(klass:, job:, e:, backoff:)
+      @remaining_retries = remaining_retries(klass, job)
```

Comment:
It's kind of nitpicky, but I'm not sure if it's ok to assign that counter to an instance variable. `QC::Worker` instance lives in the qc main process but it forks for processing each job. So while it shouldn't cause errors I would probably avoid storing `remaining_retries`.